node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+use std::{
+    collections::{HashMap, HashSet},
+    pin::Pin,
+};
+
+const COST_SIGNATURE_INVALID: ReputationChange =
+    ReputationChange::new(-100, "Bitfield signature invalid");
+const COST_MISSING_PEER_SESSION_KEY: ReputationChange =
+    ReputationChange::new(-133, "Missing peer session key");
+const COST_NOT_INTERESTED: ReputationChange =
+    ReputationChange::new(-51, "Not intersted in that parent hash");
+const COST_MESSAGE_NOT_DECODABLE: ReputationChange =
+    ReputationChange::new(-100, "Not intersted in that parent hash");
+
+
+#[derive(Default, Clone)]
+struct Tracker {
+    // track all active peers and their views
+    // to determine what is relevant to them
+    peer_views: HashMap<PeerId, View>,
+
+    // our current view
+    view: View,
+
+    // set of validators for a particular relay_parent
+    per_relay_parent: HashMap<Hash, PerRelayParentData>,
+}
+
+/// Data for each relay parent
+#[derive(Debug, Clone, Default)]
+struct PerRelayParentData {
+    // set of validators which already sent a message
+    validator_bitset_received: HashSet<ValidatorId>,
+
+    // signing context for a particular relay_parent
+    signing_context: SigningContext<Hash>,
+
+    // set of validators for a particular relay_parent
+    validator_set: Vec<ValidatorId>,
+
+    // set of validators for a particular relay_parent and the number of messages
+    // received authored by them
+    one_per_validator: HashSet<ValidatorId>,
+}
+
+fn network_update_message(n: NetworkBridgeEvent) -> AllMessages {
+    AllMessages::BitfieldDistribution(BitfieldDistributionMessage::NetworkBridgeUpdate(n))
+}
+
+pub struct BitfieldDistribution;
+
+impl BitfieldDistribution {
+    const PROTOCOL_ID: ProtocolId = *b"bitd";
+
+    async fn run<Context>(mut ctx: Context) -> SubsystemResult<()>
+    where
+        Context: SubsystemContext<Message = BitfieldDistributionMessage> + Clone,
+    {
+        // startup: register the network protocol with the bridge.
+        ctx.send_message(AllMessages::NetworkBridge(
+            NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+        ))
+        .await?;
+
+        // set of active heads the overseer told us to work on with the connected
+        // tasks abort handles
+        let mut tracker = Tracker::default();
+        loop {
+            {
+                let message = {
+                    let mut ctx = ctx.clone();
+                    ctx.recv().await?
+                };
+                match message {
+                    FromOverseer::Communication { msg } => {
+                        // we signed this bitfield
+                        match msg {
+                            // Distribute a bitfield via gossip to other validators.
+                            BitfieldDistributionMessage::DistributeBitfield(
+                                hash,
+                                signed_availability,
+                            ) => {
```

Comment:
IMO it's preferable to avoid nesting match statements:

```suggestion
                use FromOverseer::Communication;
                use BitfieldDistributionMessage::DistributeBitfield;
                
                match message {
                    // Distribute a bitfield via gossip to other validators.
                    Communication { DistributeBitfield(hash, signed_availability) } => {
```

Reply:
I am not a fan of this, since it shadows the fact that these are not structs, but enum variants.

Reply:
Yes, it obscures that they're `enum` variants, but within the context of a match arm, I'm not sure why we care. The compiler still ensures that we cover all variants, and the benefit is that each arm becomes much easier to read. 

Still, this is all a question of style, and I'm not authoritative in that respect.